### PR TITLE
Add a dpi transform to AuxTransformBox

### PR DESCRIPTION
**Pull request by has2k1**
**Issue**
While the children themselves (e.g. Text objects) of an AuxTransformBox
do conform to the dpi of the figure, the locations of the objects
do not. Therefore the space taken up by the AuxTransformBox remains
just about fixed, regardless of dpi. The relative distance between
any two objects in the AuxTransformBox remains absolutely fixed. This
leads to "scaling" artefacts for _png_ output when the dpi is not 72.

**Solution**
Add a similar dpi transformation to that done for the DrawingArea.

### CODE REVIEW

**Comment by has2k1** [comment]
Please wait on considering this PR.

**Comment by tacaswell** [comment]
That looks like it fixes a pretty major bug!

File: lib/matplotlib/offsetbox.py [hunk 1]
```diff
@@ -963,13 +966,18 @@ def get_extent(self, renderer):
         mtx = self.offset_transform.matrix_from_values(*_off)
         self.offset_transform.set_matrix(mtx)
 
-        return ub.width, ub.height, 0., 0.
+        dpi_cor = renderer.points_to_pixels(1.)
+        return ub.width*dpi_cor, ub.height*dpi_cor, 0., 0.
```
Comment (dopplershift) [inline lib/matplotlib/offsetbox.py]: Minor nit: need spaces around `*` for pep8.

Comment (has2k1) [inline lib/matplotlib/offsetbox.py]: I think pep8 is fine with that for arithmetic operators, it calls for consistency,  readability/good judgment. Either way, when I make a revision to the PR that line will not be there(for a yet known reason).

**Comment by efiring** [comment]
On 2016/10/27 1:17 PM, Hassan Kibirige wrote:
> I think pep8 is fine with that for arithmetic operators, it calls for
> consistency, readability/good judgment.

The actual PEP 8 does, exactly as you say--but the automated checker 
does not.  Contrary to the letter and the spirit of PEP 8, the automated 
checking regime is a strict application of rules.  No judgment is 
allowed.  It has its advantages in yielding consistency, but it can be 
annoying.

(Sorry, I couldn't resist the digression.)

**Comment by has2k1** [comment]
@efiring, thanks for the digression I prefer the automated solution and you have put question marks around the efficacy of the tools I use. According the `.travis.yml` the automated checker is the `pep8` package, the same thing I use for my editor. Also the manual invocation of which

```bash
pep8 lib/matplotlib/offsetbox.py | grep 970
``` 

yields nothing.

**Comment by dopplershift** [comment]
Because of course the automatic tool and the document can't be in agreement. On the plus side, I learned more about the actual document today.

File: lib/matplotlib/offsetbox.py (hunk 1 above)
Comment (NelleV) [inline lib/matplotlib/offsetbox.py]: pep8 *is* fine with that. Both the document and the tool.

**Comment by NelleV** [comment]
@dopplershift The tool and the document are on agreement on that particular point.

**Comment by dopplershift** [comment]
I *swear* I've seen it error on that--apparently I'm wrong.

**Comment by NelleV** [comment]
You are not wrong. Just outdated :-D
That particular point was fixed in a release of pep8 late 2012/early 2013

**Comment by efiring** [comment]
That's good to know--it's a big improvement.  I think that `(a*b + c*d)` is often more readable than `(a * b + c * d)`.

**Comment by NelleV** [comment]
If we want to be more reliable on enforcing pep8 on new patches, we could run flake8 on the diff of the patch instead of how we currently run pep8 (which only looks at whether the number of pep8 violations is lower or equal than the previous patch). Some projects do that (joblib, scikit-learn and maybe ipython/jupyter folks). It should be quite easy to set up.

**Comment by NelleV** [comment]
@has2k1 Do you mind explaining why we should wait on considering this PR?

**Comment by has2k1** [comment]
@NelleV, The solution so far does not completely solve the problem, as straight forward as it may be! A slight (and not good looking) modification though yields better results is not stable. I will put up a more complete illustration that should aid in getting better help.

**Comment by NelleV** [comment]
Hi @has2k1 
I know you have a better fix in the pipeline, but do you mind if we merge this in the mean time? The current patch does fix a major bug in matplotlib, and I think that having an approximate fix is much better than no fix at all.
Thanks,
N

**Comment by has2k1** [comment]
@NelleV, hold on for a day. I'll push update and a dump of my worksheet that shows the limitations.

**Comment by has2k1** [comment]
I have made the update and although sufficient for my original case, but it still does not resolve the issue.
I have not been able to come up with a simple illustration but, here is an example that illustrates the problem. (Read the text below instead of the code).

**Code**
```python
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle
from matplotlib.text import Text
from matplotlib.transforms import IdentityTransform
from matplotlib.offsetbox import (HPacker, DrawingArea, AuxTransformBox,
                                  AnchoredOffsetbox)
from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar

text = 'ghijk'
n = len(text)
xs = np.ones(n)
ys = np.arange(0, n)

width = 23
height = width*n

# "canvas" objects
da = DrawingArea(width, height, 0, 0)
auxbox = AuxTransformBox(IdentityTransform())

# patches for the "canvas". to reveal the edges
green_patch = Rectangle((0, 0), width, height,
                        fill=True, facecolor=(0, 1, 0, .3),
                        edgecolor='green')
red_patch = Rectangle((0, 0), width, height,
                      fill=True, facecolor=(1, 0, 0, .3),
                      edgecolor='red')

da.add_artist(green_patch)
auxbox.add_artist(red_patch)

# Add text to the objects
for i, (x, y, c) in enumerate(zip(xs, ys, text)):
    y = i*width
    t1 = Text(x, y, c, color='green', size=width, va='bottom')
    t2 = Text(x, y, c, color='red', size=width, va='bottom')
    da.add_artist(t1)
    auxbox.add_artist(t2)

children = [da, auxbox]
packed_box = HPacker(children=children, align='center', pad=0, sep=0)

# Figure
fig, ax = plt.subplots(figsize=(4, 6))

# boxes to be added the axes
box1 = AnchoredOffsetbox(loc=6, child=da, pad=0, frameon=True)
box2 = AnchoredOffsetbox(loc=10, child=packed_box, pad=0, frameon=True)
box3 = AnchoredOffsetbox(loc=7, child=auxbox, pad=0, frameon=True)
asb = AnchoredSizeBar(ax.transData,
                      0.5,
                      '0.5',
                      loc=8,
                      pad=0.1, borderpad=0.5, sep=5,
                      frameon=False)

ax.add_artist(box1)
ax.add_artist(box2)
ax.add_artist(box3)
ax.add_artist(asb)

fig.savefig('test-72dpi.png', dpi=72)
fig.savefig('test-144dpi.png', dpi=144)
```

*dpi=72*
test-72dpi

*dpi=144*
test-144dpi

*dpi=144, before this PR*
test-144dpi-before-patch

**Setup**
A `DrawingArea` with text, a `AuxTransformBox` with text at the same locations as the `DrawingArea`. The sizes of both boxes is defined by `Rectangle`s of the same size, green for the `DrawingArea` and `red` for the `AuxTransformBox`. The rectangles are large enough for the text to be contained within them. In the middle of the plot, the two boxes are horizontally packed. The boxes are placed onto the plot as children of `AnchoredOffsetbox` with `frameon=True`. The rectangles define the size of the boxes and shoud perfectly overlap the black frame.

Then there is the `AnchoredSizeBar` (which is implemented using an `AuxTransformBox`) at the bottom.
This is copied from the only test that uses an `AuxTransformBox`.

**Observation**
At *dpi=72*, the result is as expected. At *dpi=144* we should expect to see a "magnified" plot with all alignments in place. That is not what happens.

**Cause**
The text in the `AuxTransformBox` report larger dimensions than they should. I tried using markers, same problem.

**More info**
The culprit is deep inside the first of these lines

```python
bboxes = [c.get_window_extent(renderer) for c in self._children]
ub = mtransforms.Bbox.union(bboxes)
```

What is happening is that, since the method is called multiple times, there is a "feedback" effect that leads larger extents being reported. I was unable to counter that feedback locally in the `AuxTransformBox` class. The `Artist.get_extent`(where implemented) and `Artist.get_window_extent` methods are quite complicated in their interactions with `Artist.draw`. This made is hard to track down the source of the feedback.

**Comment by dopplershift** [comment]
Seems like `AnchoredSizeBar` got worse with this PR...

**Comment by has2k1** [comment]
@dopplershift, it seems the case but it did not. Look closely, for the original test image it is just a coincidence that the bar perfectly matches the width of the text, i.e `0.5` was the magic number for that effect.

**Comment by QuLogic** [comment]
Are you sure? The documentation says the size of the `AnchoredSizeBar` is the length in coordinates of the transform. The example specifies the data transform, so the bar should be 0.5 along the axis ticks, but it's clearly _not_.

**Comment by has2k1** [comment]
@QuLogic, you are right. I thought @doppershift meant the change in the test case, https://github.com/matplotlib/matplotlib/pull/7344/files#diff-e9a0e57a73a31b7c2a4ff11d9c338941
Rather, if reference was to  the `dpi=144` before and after, then yes.

**Comment by QuLogic** [comment]
I think the problem is that you are using a scaling Affine 2D transform for the DPI correction, which scales both location _and_ size, but the size is already correct because the font is twice as big. So the position of the "k" child is where you want, but its calculated size is twice as big, so that's why you get what appears to be an extra line above it in the frame.

What you need to apply for the DPI transform is something that will only scale the position of children. Is there any easy way to do this? I'm not sure, transforms are tricky...

**Comment by QuLogic** [comment]
I had an idea while I was sleeping; since `AuxTransformBox` draws each child itself, the transform doesn't have to be strictly correct as long as it can be adjusted correctly. Adding the correct DPI scaling to _just_ the position works pretty well for the text:
test-144dpi
However, the `Rectangle` will not be scaled in size at the same time. We could special one or the other, but I feel that's becoming a bit of a hack.

Based on how `AnchoredSizeBar` works with `AuxTransformBox`, I'm 50/50 on whether it's a bug or not. In the example, an identity transform is used, which naturally does not account for DPI, so it really wrong that the positions don't change? I don't know; I've been flip-flopping on whether it is or isn't.

**Comment by has2k1** [comment]
@Qulogic
Indeed, how `AnchoredSizeBar` works makes this a more stubborn issue. I am tempted to recommend rewriting it if there is an alternate way, maybe using a `Line2D` do.

There are a bunch of permutations for trying to isolate the scaling and you have hit on results I do not recall having seen. I have encountered the difference in scaling between the `text` and the `Rectangle`.  Creating special cases is definitely not an appealing solution.

A better solution I have hit upon is to not clear the DPI transform and let the children compute the extents with it. So there is no need to multiply the sizes of the Box by the DPI correction. I think this is related to your earlier comment. However, it breaks `AnchoredSizeBar` not really sure why.

test-144dpi

If that holds for all kinds of objects put inside the box, then the case for changing the implementation of `AnchorSizeBar` becomes stronger.

On a larger note while tussling with how the extents are computed across various `Artist`s, I am filled with angst about the feasibility of implementing a constraint based layout manager.

**Comment by QuLogic** [comment]
Try enabling the frame on the `AnchoredSizeBar` and see if you can figure out where it really is now.

**Comment by has2k1** [comment]
I did that. It is at the bottom. Some of it below the x-axis.

**Comment by has2k1** [comment]
My previous response was facilitated by a poor memory. This is the result.
test-144dpi

The problem is that for the `AnchoredSizeBar` the `AuxTransformBox` is created with `transform=ax.transData`. If you used an `IdentityTransform` as is the case for the red boxes, then it works.

**Comment by efiring** [comment]
Is this ready to go after a rebase?  Or are there still problems with it?  If so, is it *really* release-critical?

**Comment by QuLogic** [comment]
I don't think this is one rebase away from being ready; there's still the question of what to do with `AnchoredSizeBar`. I've convinced myself that it is working correctly, so something will have to be done to ensure it works the same way once this PR is okay.

You'll have to ask @NelleV why it's release critical.

**Comment by has2k1** [comment]
I am in agreement with @QuLogic. I have without success tried to find a fix for `AnchoredsizeBar`.

**Comment by NelleV** [comment]
We can remove the release critical label, but this is a *major* bug. All the publications that use this are wrong, and we are probably talking thousands of publications.

**Comment by QuLogic** [comment]
I'm not so sure it's that major. `AuxTransformBox` is not used anywhere in matplotlib proper. It is used in `axes_grid1.anchored_artists.{AnchoredAuxTransformBox,AnchoredEllipse,AnchoredSizeBar}` and provided as an alias in `axes_grid1.anchored_artists`. In the former case, the documentation suggests using something like `ax.transData` as the transform, which includes DPI automatically. `AuxTransformBox` itself is not really well documented.

This bug only arises when using `AuxTransformBox` with `IdentityTransform`, which by definition does not include a DPI adjustment, _and_ it only affects text because fonts are always specified in points which do obey DPI. So is it really a bug, or is just an unfortunate combination of many factors? Are there actually a whole lot of publications using `AuxTransformBox` in this manner, or do they use one of the wrappers, which seem to work fine if used as documented?

**Comment by QuLogic** [comment]
I'm un-critical-izing this as I don't think it will be fixed soon and I'm not so sure it's that critical.

**Comment by jklymak** [comment]
I'm pretty confused if this really needs to be fixed or just closed.  Hopefully one of the original proponents can chime in again?  Otherwise I've marked stale and will close in a month or so.

**Comment by has2k1** [comment]
As I recall, it is not easy fix this issue without breaking `AnchoredSizeBar` which relies on the "bug". If we should to fix this, we can create a special (internal) subclass of `AuxTransformBox` that is used by `AnchoredSizeBar` so that it is not broken by the solution.

For my use case in Plotnine, I created a [AuxTransformBox subclass](https://github.com/has2k1/plotnine/blob/469de2629e5c930f46e682dd369ec6ba722884f8/plotnine/guides/guide_colorbar.py#L393-L421) that worked as I expected.

**Comment by jklymak** [comment]
I don't at all mind this going forward...  I've just moved to draft, and feel free to update or start again if you are interested.  I'm just cleaning up and haven't thought about it deeply.

**Comment by jklymak** [comment]
I'll close as abandoned here.

**Pull request closed without merging**